### PR TITLE
Fixes #22372 - Audit - Sync Plans, Product, GPG & Activation Keys, HostCollection

### PR DESCRIPTION
**Pull request by kgaikwad**

### CODE REVIEW

**Comment by jlsherrill** [comment]
@kgaikwad  looks like there are some test failures with this.

**Comment by kgaikwad** [comment]
@ares & @jlsherrill ,

Fixed the test failures for `Audit` from file [association_test.rb](https://github.com/Katello/katello/blob/master/test/models/association_test.rb) . 
PR is ready for review.

**Comment by ares** [comment]
code looks good to me, couple notes from tesing:

1) when I upload manifest, I see many Katello/Products audit entries created by anonymous admin, I think we should ignore such creation as it's done by the system, I think we need to add `Katello::Product.ignore_auditing do ... end` around this code https://github.com/Katello/katello/blob/d561b9f77312ff3fe20687daea0fe5cc2a357dc7/app/lib/actions/candlepin/owner/import_products.rb#L11

2) if I create Content credential which is not GPG key but SSL certificate, audit still lists it as GPG. We should probably override `to_label` of this class to content credential

3) creating Activation Key creates 2 audits, one for the creation, then second for update saying "Cp changed from N/A to N/A", we should ignore :cp_id attribute (marked in code)

4) when I assign AK to some product, I see following audit "Gpg key changed from N/A to N/A", looking at changes in DB I see this
```
Audit.last.audited_changes
=> {"gpg_key_id"=>[nil, 1]}
```
not sure why the key id is not properly printed, the same issue was for cp_id in 3)

Otherwise it looks as great improvement already!

**Comment by kgaikwad** [comment]
@ares ,

Thank you for testing and sharing your observations about changes.

> 1. when I upload manifest, I see many Katello/Products audit entries created by anonymous admin, I think we should ignore such creation as it's done by the system, I think we need to add Katello::Product.ignore_auditing do ... end around this code

Done.

> 2. if I create Content credential which is not GPG key but SSL certificate, audit still lists it as GPG. We should probably override to_label of this class to content credential

overridden to_label of GPG key to return value `content credential (#{content_type} - #{name})`
Please confirm about value for to_label.

screenshot from audit 2018-02-28 12-47-13

> 3. creating Activation Key creates 2 audits, one for the creation, then second for update saying "Cp changed from N/A to N/A", we should ignore :cp_id attribute (marked in code)

Ignored :cp_id attribute

> 4. when I assign AK to some product, I see following audit "Gpg key changed from N/A to N/A", looking at changes in DB I see this

Assuming that you wanted to mention Gpg key instead of AK.
Yes, I debugged on this `N/A to N/A`.
Found that I will need to add changes related to Katello here in [audits_helper.rb](https://github.com/theforeman/foreman/blob/develop/app/helpers/audits_helper.rb#L6)


Updated PR as per suggested changes. Ready for review.

**Comment by ares** [comment]
code looks good to me, @rplevka mind to test too?

**Comment by rplevka** [comment]
@kgaikwad hi :)

- Whenever I create any katello entity I can't see details in the message. all i get is "my_syncplan_1 to N/A". Wouldn't it be beneficial if we logged some important attributes on create? (e.g. sync dates on sync plan, etc.)
- I kind of miss the katello auditable_types when i try to search by type - anyway, i'm not sure whether this PR is supposed to address that:
```
auditable_type' should be one of 'architecture, auth_source, bookmark, compute_attribute, compute_profile, compute_resource, config_group, domain, host, hostgroup, image, interface, location, medium, os, organization, override_value, partition_table, parameter, puppetclass, realm, role, setting, ssh_key, smart_proxy, subnet, user, usergroup, template, provisioning_template, ptable'
```
- activation-key don't record any log if a user assign/unassign a subscription to/from it
- activation-key don't record any log if a user assign/unassign a host collection to/from it 

- katello-related audit records use only the "info" label syling (they're all blue):
screenshot


that's all i've got for now, thanks :)

**Comment by kgaikwad** [comment]
@rplevka,

All the points apart from `activation-key` are mainly related to audit list page. 
To fix these above points, I need to override existing audit helper files from foreman into Katello. 


```
Whenever I create any katello entity I can't see details in the message. all i get is "my_syncplan_1 to N/A". Wouldn't it be beneficial if we logged some important attributes on create? (e.g. sync dates on sync plan, etc.)
```
While working on audit helpers, I have observed that we are not showing any message for object creation in case of foreman resources. After adding few changes to audit helper file, now it is not showing any message under audit record on web-ui not even - "my_syncplan_1 to N/A".
@ares, is that correct or am I missing anything?
Currently, I have added [helper extension changes](https://github.com/kgaikwad/katello/commit/3ea64e6b588002ef22d027811948e19c24edae79) but I think I need to add more changes related to searching using [audit_extensions.rb](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/audit_extensions.rb)

```
- activation-key don't record any log if a user assign/unassign a subscription to/from it
- activation-key don't record any log if a user assign/unassign a host collection to/from it

```
About `activation-key` auditing including few other Katello resources like Host Collections, GPG Keys, Product there are no specifications mentioned in google doc so I have covered only CRUD actions for these resources.

@ares, please confirm if we are going to audit above associations with `activation-key` so that I will go & add changes related to `activation-key` based on that.

**Comment by kgaikwad** [comment]
@ares & @rplevka ,

Moved `audited` method at the top in models. 

I have covered audit changes -  `activation-key don't record any log if a user assign/unassign a host collection to/from it` in [PR-7248](https://github.com/Katello/katello/pull/7248)

I will confirm and create separate ticket for ` - activation-key don't record any log if a user assign/unassign a subscription to/from it`.

Remaining suggested changes are related to web-ui.
so if changes are working as expected then we can merge this PR.

File: app/lib/actions/candlepin/owner/import_products.rb [hunk 1]
```diff
@@ -8,8 +8,10 @@ class ImportProducts < Candlepin::Abstract
 
         def run
           organization = ::Organization.find(input[:organization_id])
-          User.as_anonymous_admin do
-            organization.redhat_provider.import_products_from_cp
+          Katello::Product.ignore_auditing do
```
[the remark this answers is not in the document]
Comment (kgaikwad) [reply app/lib/actions/candlepin/owner/import_products.rb]: @ares, 
Sorry, my bad.
I will fix this. Apologize! I have referred your comment and I didn't test it properly. I should have test this scenario.

**Comment by ares** [comment]
do you find missing information if you navigate to audit detail? (clicking on prod2)

**Comment by kgaikwad** [comment]
@rplevka,

Thank you for testing this PR again.

> assigning repos to a sync plan generates 2 identical audit messages - also, the message is quite useless - also, one is unable to tell whether a product got assigned or removed from the sync plan

Debugging on this. Removing a product from sync-plan creates only single entry in audit but adding product to sync-plans creates two identical entries in audit. 
Currently, it is showing N/A to N/A because we will need to extend some core helper files.
Then, it will show actual values. But as per my understanding message will be same in that case as well.
@ares, please correct me if misunderstood somewhere.

> creating some instance (e.g. product) with a description (and probably any other parameter) generates kind of confusing messages:

Again, extending core helper files will fix this. After clicking product, I can see all information except `Provider`.

I have made few changes on my local related to this MAIN_OBJECTS. Please check below images.

Before changes: 
audit-prod-without-main-objects

After few changes I made to MAIN_OBJECTS and few helper methods on local instance:
audit-prod-with-main-objects

**Comment by ares** [comment]
I tested and was able to reproduce. To avoid double auditing and properly audit the association between product and sync plan I think following should be applied

```diff
diff --git a/app/models/katello/sync_plan.rb b/app/models/katello/sync_plan.rb
index 613400b..cbf9bd3 100644
--- a/app/models/katello/sync_plan.rb
+++ b/app/models/katello/sync_plan.rb
@@ -3,7 +3,7 @@ require 'time'
 module Katello
   class SyncPlan < Katello::Model
     audited
-    has_associated_audits
+    audit_associations :products
     include Glue
     include Katello::Authorization::SyncPlan
     include ForemanTasks::Concerns::ActionSubject
```

The problem with N/A must be fixed in core for which I opened a PR https://github.com/theforeman/foreman/pull/5449

**Comment by kgaikwad** [comment]
@ares, 
Thank you! Updated the pull-request with suggested changes.

**Comment by rplevka** [comment]
@ares, @kgaikwad - hi again. I tested the GPG keys and it seems that the changed content is not redacted. is it okay that an auditor sees a complete key? I think it is not.

**Comment by rplevka** [comment]
@ares 
ok. thanks for clarifying. the value is truncated in the ui, however it's fully visible using /api/audits endpoint

File: app/models/katello/gpg_key.rb [hunk 2]
```diff
@@ -1,5 +1,8 @@
 module Katello
   class GpgKey < Katello::Model
+    audited
+    audit_associations :products
```
Comment (tbrisker) [inline app/models/katello/gpg_key.rb]: you can combine both in one as `audited :associations => :products`, `audit_associations` is mostly needed when we want to add associations to an existing model.

File: app/models/katello/sync_plan.rb [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 module Katello
   class SyncPlan < Katello::Model
+    audited
+    audit_associations :products
```
Comment (tbrisker) [inline app/models/katello/sync_plan.rb]: ditto

**Comment by kgaikwad** [comment]
@tbrisker,

Thank you!
With `audited` & `audit_associations` change fixed the association but was still creating duplicate audit records for products as @rplevka mentioned above.
Removing it to `audited :associations => [:products]` fixes both the issues properly.

Updated the pull-request as per suggested changes.

**Comment by ares** [comment]
I did another round of review together with #7248, it works fine for listed resources. I think this can be merged now.

**Comment by bbuckingham** [comment]
@kgaikwad, Thanks for the PR!  

Based on discussion with @ares, I am going to proceed with merging.

**Pull request merged**